test/test_python_dispatch.py

```diff
@@ -517,11 +514,82 @@ def test_enable_python_mode_respects_no_dispatch(self) -> None:
                 self.assertEqual(z.elem, expected)
 
     def test_nested_enable_python_mode(self) -> None:
-        with self.assertRaisesRegex(RuntimeError, "has already been set"):
+        class ErrorA(RuntimeError):
+            pass
+
+        class A(torch.Tensor):
+            @staticmethod
+            def __new__(cls, elem):
+                return torch.Tensor._make_subclass(cls, elem, elem.requires_grad)
+
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+                raise ErrorA
+
+        with self.assertRaisesRegex(ValueError, "there is already an active mode"):
             with enable_python_mode(LoggingTensor):
-                with enable_python_mode(LoggingTensor):
+                with enable_python_mode(A):
                     pass
 
+        # nested enable_python_modes are allowed if they're the same mode
+        # Will only write once to the log
+        with capture_logs() as logs:
+            x = LoggingTensor(torch.tensor([3.]))
+            log_input("x", x)
+            with enable_python_mode(LoggingTensor):
+                with enable_python_mode(LoggingTensor):
+                    x + x
```

Comment:
Does torch_function mode have the same behavior?

I'm not sure how I feel about this. It's clearer if both things passed in to enable_python_mode are actually Mode objects, because then if two different mode instances are passed in then that counts as nesting mode but if it is the same mode instance then this behavior seems intuitive

Reply:
Yes. I'll split this test function so it's clearer. Nesting does _not_ work with `enable_python_mode` generally (this is tested in the lines above), but there's a special case where it will ignore the second context manager if the mode/tensor subclass passed to it is the same